### ISSUE TITLE
Can't compile program statically

### ISSUE THREAD

**Issue by inem** [crystal-lang/crystal#2003]
I want to share executable, and therefore I want to compile my crystal program statically.

It looks like it needs these libs:
```
/usr/local/opt/libevent/lib/libevent-2.0.5.dylib
/usr/local/opt/libpcl/lib/libpcl.1.dylib
/usr/local/opt/bdw-gc/lib/libgc.1.dylib
```

But when I try to compile it I get:
```bash
crystal build activity.cr --link-flags "-static -L/usr/local/opt/libevent/lib/" 
ld: library not found for -lcrt0.o
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Error: execution of command failed with code: 1: `/usr/bin/gcc -o "activity" "${@}" -static -L/usr/local/opt/libevent/lib/ -rdynamic  -levent -lpcl -lpcre -lgc -lpthread -ldl`
```

Does anybody have an idea what am I doing wrong and how to do it properly?

I am on OSX 10.11, Crystal v0.10.0

**Comment by ysbaddaden** [comment]
This is an OSX problem. You can't fully statically compile an executable.

You can however have the libevent.a, libgc a and libpcl.a be linked statically, but not all the system OSX libs. But you must trick the linker to only link the .a (it chooses .dylib when it finds them). You may include --link-flags "path/to/libevent.a ..." or change LIBRARY_PATH or some other trick.

**Comment by inem** [comment]
Thanks, I was thinking about to try similar solution.

> This is an OSX problem.
Can I create executable for OS X under Linux?

**Comment by inem** [comment]
@ysbaddaden I tried your advice, and got it compiled without any issues:
`crystal build activity.cr --link-flags "/usr/local/opt/libevent/lib/libevent.a /usr/local/opt/libpcl/lib/libpcl.a /usr/local/opt/bdw-gc/lib/libgc.a"`

And size of the binary is now bigger.

Though on another system, I still get the same error:
<img width="430" alt="c3giyetc_png__847x299__and_bpx_-_subscription_system_prototype" src="https://cloud.githubusercontent.com/assets/9665/12221154/bcae66ec-b78e-11e5-9ba1-20305cf60b5a.png">

**Comment by ysbaddaden** [comment]
The problem is if a .dylib file can be found, the linker will always use it over .a files (the same happens on linux with .so files).

I'm not a veteran C programmer, and neither an OSX user, so I can't help much more. I only remember I struggle to make semi static binaries on OSX (and quickly put the solution that worked for me in a Makefile).

**Comment by inem** [comment]
Doesn't seem to work anyway.
It ends up asking for `lcrt0.o`, and there is [no such file on OS X](http://stackoverflow.com/questions/844819/how-to-static-link-on-os-x).

So looks like it is impossible.

**Comment by ysbaddaden** [comment]
I somehow achieved it with Shards, but yes, not a real static binary.

**Comment by luislavena** [comment]
@inem OSX and all official compilers in that platform discourage and disable functionality that will allow you to produce *true* static binaries.

https://developer.apple.com/library/mac/qa/qa1118/_index.html

Please note that I said *binaries*, but is still possible to generate static libraries.

I don't recall if the linker (ld) provided by OSX supports `-Bstatic` and `-Bdynamic` but playing with that should allow you to link to the static version of libraries while retain the dynamic system libraries that OSX imposes on you.

You can call this via `cc` using `-Wl,-Bstatic -lyaml -lmylib -Wl,-Bdynamic`

That will ensure *just* the libraries between `static` and `dynamic` are linked statically.

I no longer have access to an OSX platform so you will need to fiddle with these instructions.

Hope that helps.

**Comment by luislavena** [comment]
Sorry, forgot to mention another approach, which is documented here:

- https://developer.apple.com/library/mac/qa/qa1393/_index.html
- http://stackoverflow.com/a/4576798/117298

Cheers.

**Comment by jwoertink** [comment]
Hey, I'm new to all this, and just want to make sure I'm understanding correctly here. I installed crystal, wrote a program

```crystal
require "colorize"
puts "THIS SHOULD BE GREEN".colorize(:green)
```
Then ran `crystal build sample.cr --release`, and sent that `sample` program to a friend running OSX that does not have crystal installed. When he ran `./sample`, he got

```text
➜  Downloads  ./sample
dyld: Library not loaded: /usr/local/opt/bdw-gc/lib/libgc.1.dylib
  Referenced from: /Users/russ/Downloads/./sample
  Reason: image not found
[1]    6606 trace trap  ./sample
```

So to release these programs to people without crystal, I need to link this `libgc.1.dylib` file? I'm running *0.15.0* on OSX 10.11.4

**Comment by luislavena** [comment]
@jwoertink as stated above, OSX is not very fond to static linking. Also, when you compiled your program, you didn't add the linking flags to indicate that you want (or prefer) static libraries (see comments above about using `--link-flags`).

When using `--link-flags="-static"` it should try to link the binary against the static version of available libraries.

Of course, it might fail on OSX since static linking is not supported (as mentioned above), so is more a try and error (and more errors than tries).

Hope that helps.

**Comment by ysbaddaden** [comment]
@luislavena we could consider that it should be easy to share/distribute an OSX binary. Despite being required, gc, event and pcre can be considered implementation details to Crystal.

Debian/Ubuntu packages don't have this problem because static libraries are distributed along with the compiler and linking will use those static libraries (unless the development packages are installed). This isn't the case for the homebrew crystal, which relies on installed homebrew libraries, where both static and shared libraries are installed, and shared always takes precedence.

@jwoertink you can try to tell the linker to use the static libraries when building your app:

```sh
crystal build myapp.cr --link-flags "/usr/local/lib/libgc.a"
```

You can specify has many libraries are necessary like `--link-flags "/usr/local/lib/libgc.a /usr/local/lib/libevent_core.a"`.

**Comment by jwoertink** [comment]
@ysbaddaden How do I know what needs to be linked? In the case of my example, I'm doing a next level "hello world" type example. 

I tried doing `crystal build sample.cr --link-flags "/usr/local/lib/libgc.a"`, but running `./sample` from the other machine gives the same error. I saw that Apple supports "Static library" but not "statically linked binary". I'm still not sure exactly what that means, but in any case, I still totally dig this project :smile:
